### PR TITLE
integrate ngfx backend + various node.gl optimizations

### PR DESCRIPTION
**Pull request by jmoguillansky-gpsw**

### CODE REVIEW

**Comment by cboesch-gpsw** [comment]
A few obvious things before we can start looking into this:
1. "This branch has conflicts that must be resolved" → please rebase on vulkan
2. please remove your TODO about downloading; as I already said, it's not acceptable to have external code in the tree, even in a dev branch. One of the main reason is that it makes the repo larger for everyone.
3. It looks like you are shuffling the whole structure all around; we never discussed that and I though the consensus was to inline the backends instead of having ngfx

**Comment by jmoguillansky-gpsw** [comment]
please remove your TODO about downloading; as I already said, it's not acceptable to have external code in the tree, even in a dev branch. One of the main reason is that it makes the repo much larger for everyone.

>> most of the external dependencies are downloaded automatically.  Only ffmpeg binary is included and I will update so that it's downloaded automatically.  

 It looks like you are shuffling the whole tree structure all around; we never discussed that and I though the consensus was to inline the backends instead of having ngfx

>> There's two motivations for shuffling nodegl folder structure.  First: to resolve build issues.  When we do #include "format.h" for example it's ambiguous whether we're referring to system format.h, an external package "format.h" or node.gl format.h.  
The second motivation is so that the code is easier to manage.  The build file is simplified significantly.  Instead of specifying all the filenames specified in the build system, we can just specify to include all the files in a folder using a glob operator.  
Feel free to reorganize the folder structure as you wish, but it would be helpful if we have some sort of folder structure instead of including nodegl headers in a global space, to avoid name collisions. 

With regards to ngfx and shader-tools module and how to organize the files, feel free to propose a change.

**Comment by jmoguillansky-gpsw** [comment]
"I though the consensus was to inline the backends instead of having ngfx" : can you clarify this comment?
all the ngfx integration code is in libnodegl/src/nodegl/porting/ngfx/
the ngfx module code itself is currently in top level folder but we can move it if you like, to external folder, etc.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    
+if (${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+
+set (PTHREADS_DIR ${EXTERNAL_DIR}/win64/pthreads-w32-2-9-1-release/Pre-built.2)
+set (PTHREADS_INCLUDE_DIRS ${PTHREADS_DIR}/include)
+set (PTHREADS_CFLAGS -DHAVE_STRUCT_TIMESPEC)
+set (PTHREADS_LIBRARY_DIRS ${PTHREADS_DIR}/lib/x64)
+set (PTHREADS_LIBRARIES pthreadVC2)
+
+set (FFMPEG_DIR ${EXTERNAL_DIR}/win64/ffmpeg-20200831-4a11a6f-win64-dev)
```
Comment (jmoguillansky-gpsw) [inline CMakeLists.txt]: TODO: rebase

File: CMakeLists.txt [hunk 2]
```diff
@@ -19,7 +19,7 @@ set (PTHREADS_CFLAGS -DHAVE_STRUCT_TIMESPEC)
 set (PTHREADS_LIBRARY_DIRS ${PTHREADS_DIR}/lib/x64)
 set (PTHREADS_LIBRARIES pthreadVC2)
 
-set (FFMPEG_DIR ${EXTERNAL_DIR}/win64/ffmpeg-20200831-4a11a6f-win64-dev)
+set (FFMPEG_DIR ${EXTERNAL_DIR}/win64/ffmpeg-4.3.1-2020-09-21-full_build-shared)
```
Comment (jmoguillansky-gpsw) [inline CMakeLists.txt:22]: TODO: update previous CMakeFiles commit

File: libnodegl/src/nodegl/core/buffer.h [hunk 3]
```diff
@@ -42,7 +43,10 @@ struct buffer {
 
 struct buffer *ngli_buffer_create(struct gctx *gctx);
 int ngli_buffer_init(struct buffer *s, int size, int usage);
-int ngli_buffer_upload(struct buffer *s, const void *data, int size);
+int ngli_buffer_upload(struct buffer *s, const void *data, uint32_t size, uint32_t offset);
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/core/buffer.h:46]: TODO: update history.  move offset param to separate commit

File: libnodegl/src/nodegl/porting/vk/buffer_vk.c [hunk 4]
```diff
@@ -60,7 +60,7 @@ int ngli_buffer_vk_init(struct buffer *s, int size, int usage)
     };
     VkResult res = vkCreateBuffer(vk->device, &buffer_create_info, NULL, &s_priv->buffer);
     if (res != VK_SUCCESS)
-        return -1;
+        return NGL_ERROR_GENERIC;
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/porting/vk/buffer_vk.c:63]: TODO: move to separate commit

File: libnodegl/src/nodegl/core/hwconv.c [hunk 5]
```diff
@@ -107,7 +107,7 @@ int ngli_hwconv_init(struct hwconv *hwconv, struct ngl_ctx *ctx,
     if (ret < 0)
         return ret;
 
-    ret = ngli_buffer_upload(hwconv->vertices, vertices, sizeof(vertices));
+    ret = ngli_buffer_upload(hwconv->vertices, vertices, sizeof(vertices), 0);
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/core/hwconv.c:110]: TODO: update history.  add changes to buffer.c/h

File: libnodegl/src/nodegl/core/pgcraft.c [hunk 6]
```diff
@@ -442,7 +445,12 @@ static int inject_block(struct pgcraft *s, struct bstr *b,
     }
 
     const char *keyword = get_glsl_type(block->type);
-    ngli_bstr_printf(b, " %s %s_block {\n", keyword, named_block->name);
+    struct bstr *memory_qualifiers = ngli_bstr_create();
+    if (config->backend != NGL_BACKEND_OPENGL) {
+        if ((0 == strcmp(keyword, "buffer")) &&  (stage != NGLI_PROGRAM_SHADER_COMP)) ngli_bstr_print(memory_qualifiers , "readonly");
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/core/pgcraft.c:450]: The gl compiler reported a compile error when including keyword "readonly"

File: libnodegl/src/nodegl/core/rendertarget.c [hunk 7]
```diff
@@ -42,6 +42,16 @@ void ngli_rendertarget_read_pixels(struct rendertarget *s, uint8_t *data)
     s->gctx->clazz->rendertarget_read_pixels(s, data);
 }
 
+void ngli_rendertarget_begin_pass(struct rendertarget *s)
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/core/rendertarget.c:45]: TODO: update history.  merge with previous commit

File: libnodegl/src/nodegl/porting/egl/glcontext_egl.c [hunk 8]
```diff
@@ -327,8 +327,9 @@ try_again:;
 
             if (i)
                 LOG(WARNING, "falling back on OpenGL %d.%d", gl_versions[i].major, gl_versions[i].minor);
-
             egl->handle = eglCreateContext(egl->display, egl->config, shared_context, ctx_attribs);
+            extern void* _capture_device;
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/porting/egl/glcontext_egl.c:331]: TODO: remove this global variable.  This is a hack for debugging

File: libnodegl/src/nodegl/porting/gl/glfeatures_data.h [hunk 9]
```diff
@@ -251,5 +251,13 @@ static const struct glfeature {
                                            OFFSET(Uniform3uiv),
                                            OFFSET(Uniform4uiv),
                                            -1}
+    }, {
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/porting/gl/glfeatures_data.h:254]: TODO: update history.  merge this commit with previous commit

File: libnodegl/src/nodegl/porting/vk/gctx_vk.c [hunk 10]
```diff
@@ -453,35 +453,6 @@ static int vk_resize(struct gctx *s, int width, int height, const int *viewport)
     return 0;
 }
 
-static int vk_swap_buffers(struct gctx *s)
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/porting/vk/gctx_vk.c]: TODO: update history.  this portion should be part of previous commit

File: libnodegl/src/nodegl/core/pgcraft.c [hunk 11]
```diff
@@ -222,7 +222,7 @@ static const int texture_types_map_gl[NGLI_PGCRAFT_SHADER_TEX_TYPE_NB][NGLI_INFO
         [NGLI_INFO_FIELD_Y_SAMPLER]         = NGLI_TYPE_SAMPLER_2D,
         [NGLI_INFO_FIELD_UV_SAMPLER]        = NGLI_TYPE_SAMPLER_2D,
         [NGLI_INFO_FIELD_COLOR_MATRIX]      = NGLI_TYPE_MAT4,
-#elif defined(TARGET_DARWIN)
+#elif (defined(TARGET_DARWIN) && defined(BACKEND_GL))
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/core/pgcraft.c:225]: This should be if (config->gl) to check current backend

File: libnodegl/src/nodegl/core/log.c [hunk 12]
```diff
@@ -22,8 +22,9 @@
 #include <stdarg.h>
 #include <stdio.h>
 #include <stdlib.h>
+#ifndef WIN32
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/core/log.c:25]: TODO: add comment in commit message: unistd not support on windows with VS toolchain

File: libnodegl/src/nodegl/core/memory.c [hunk 13]
```diff
@@ -79,17 +80,15 @@ void *ngli_calloc(size_t n, size_t size)
 
 void *ngli_malloc_aligned(size_t size)
 {
-    if (failure_requested())
-        return NULL;
-
-    void *ptr;
-#ifdef TARGET_MINGW_W64
-    ptr = _aligned_malloc(size, NGLI_ALIGN_VAL);
+#if defined(WIN32)
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/core/memory.c:83]: TODO: on windows use aligned_malloc. On other platforms keep current code

File: libnodegl/src/nodegl/core/pgcraft.c [hunk 14]
```diff
@@ -250,7 +250,7 @@ static const int texture_types_map_default[NGLI_PGCRAFT_SHADER_TEX_TYPE_NB][NGLI
         [NGLI_INFO_FIELD_COORDINATE_MATRIX] = NGLI_TYPE_MAT4,
         [NGLI_INFO_FIELD_DIMENSIONS]        = NGLI_TYPE_VEC2,
         [NGLI_INFO_FIELD_TIMESTAMP]         = NGLI_TYPE_FLOAT,
-#if defined(TARGET_DARWIN) || defined(TARGET_IPHONE)
+#if (defined(TARGET_DARWIN) || defined(TARGET_IPHONE)) && defined(BACKEND_GL)
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/core/pgcraft.c:253]: TODO: don't use macro defined(BACKEND_GL), use config

File: libnodegl/src/nodegl/porting/gl/glincludes.h [hunk 15]
```diff
@@ -53,9 +53,11 @@
 #endif
 
 #if _WIN32
-# include <GL/gl.h>
-# include <GL/glcorearb.h>
-# include <GL/glext.h>
+#include <windows.h>
+#include <GL/glew.h>
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/porting/gl/glincludes.h:57]: TODO: investigate can we avoid glew, add mingw support

**Review by jmoguillansky-gpsw — COMMENTED** [review]
.

**Review by jmoguillansky-gpsw — COMMENTED** [review]
.

File: libnodegl/src/nodegl/core/animation.h [hunk 16]
```diff
@@ -24,6 +24,10 @@
 
 #include "nodegl.h"
 
+#ifdef __cplusplus
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/core/animation.h:27]: Motivation:
simplifies overall code. Instead of putting extern "C" in every file that includes a node.gl header file from C++ code, we can put extern "C" once in the node.gl header file and it can be included from either C or C++ code

File: pynodegl-utils/pynodegl_utils/examples/shaders/color.vert [hunk 17]
```diff
@@ -1,4 +1,4 @@
 void main()
 {
-    ngl_out_pos = ngl_projection_matrix * ngl_modelview_matrix * ngl_position;
+    ngl_out_pos = ngl_projection_matrix * ngl_modelview_matrix * vec4(ngl_position, 1.0);
```
Comment (jmoguillansky-gpsw) [inline pynodegl-utils/pynodegl_utils/examples/shaders/color.vert:3]: motivation: the glsl input type matches the node graph data type.  This is also more aligned with current pgcraft architecture, where input type is inferred from node graph

File: libnodegl/src/nodegl/core/buffer.c [hunk 18]
```diff
@@ -26,22 +26,22 @@
 
 struct buffer *ngli_buffer_create(struct gctx *gctx)
 {
-    return gctx->class->buffer_create(gctx);
+    return gctx->clazz->buffer_create(gctx);
```
Comment (jmoguillansky-gpsw) [inline libnodegl/src/nodegl/core/buffer.c:29]: Motivations: 1: using keyword "class" as a variable name can be confusing to a C++ developer, 2: IDEs may be confused by "class", 3: code analysis tools / style tools may be confused by keyword "class" as a variable name

**Review by jmoguillansky-gpsw — COMMENTED** [review]
" nodegl: change folder structure ":

here's corresponding commands:
mkdir -p libnodegl/src/nodegl/core
mkdir -p libnodegl/src/nodegl/porting/android
mkdir -p libnodegl/src/nodegl/porting/eagl
mkdir -p libnodegl/src/nodegl/porting/egl
mkdir -p libnodegl/src/nodegl/porting/gl
mkdir -p libnodegl/src/nodegl/porting/hwupload
mkdir -p libnodegl/src/nodegl/porting/ngfx
mkdir -p libnodegl/src/nodegl/porting/nsgl
mkdir -p libnodegl/src/nodegl/porting/vaapi
mkdir -p libnodegl/src/nodegl/porting/vk
mkdir -p libnodegl/src/nodegl/porting/wgl
mkdir -p libnodegl/test
mkdir -p libnodegl/tools/gen_doc
mkdir -p libnodegl/tools/gen_specs
mv libnodegl/animation.c libnodegl/src/nodegl/core/animation.c
mv libnodegl/animation.h libnodegl/src/nodegl/core/animation.h
mv libnodegl/api.c libnodegl/src/nodegl/core/api.c
mv libnodegl/asm_aarch64.S libnodegl/src/nodegl/core/asm_aarch64.S
mv libnodegl/block.c libnodegl/src/nodegl/core/block.c
mv libnodegl/block.h libnodegl/src/nodegl/core/block.h
mv libnodegl/bstr.c libnodegl/src/nodegl/core/bstr.c
mv libnodegl/bstr.h libnodegl/src/nodegl/core/bstr.h
mv libnodegl/buffer.c libnodegl/src/nodegl/core/buffer.c
mv libnodegl/buffer.h libnodegl/src/nodegl/core/buffer.h
mv libnodegl/colorconv.c libnodegl/src/nodegl/core/colorconv.c
mv libnodegl/colorconv.h libnodegl/src/nodegl/core/colorconv.h
mv libnodegl/darray.c libnodegl/src/nodegl/core/darray.c
mv libnodegl/darray.h libnodegl/src/nodegl/core/darray.h
mv libnodegl/deserialize.c libnodegl/src/nodegl/core/deserialize.c
mv libnodegl/dot.c libnodegl/src/nodegl/core/dot.c
mv libnodegl/drawutils.c libnodegl/src/nodegl/core/drawutils.c
mv libnodegl/drawutils.h libnodegl/src/nodegl/core/drawutils.h
mv libnodegl/egl.h libnodegl/src/nodegl/core/egl.h
mv libnodegl/features.h libnodegl/src/nodegl/core/features.h
mv libnodegl/format.c libnodegl/src/nodegl/core/format.c
mv libnodegl/format.h libnodegl/src/nodegl/core/format.h
mv libnodegl/gctx.c libnodegl/src/nodegl/core/gctx.c
mv libnodegl/gctx.h libnodegl/src/nodegl/core/gctx.h
mv libnodegl/graphicstate.h libnodegl/src/nodegl/core/graphicstate.h
mv libnodegl/gtimer.c libnodegl/src/nodegl/core/gtimer.c
mv libnodegl/gtimer.h libnodegl/src/nodegl/core/gtimer.h
mv libnodegl/hmap.c libnodegl/src/nodegl/core/hmap.c
mv libnodegl/hmap.h libnodegl/src/nodegl/core/hmap.h
mv libnodegl/hwconv.c libnodegl/src/nodegl/core/hwconv.c
mv libnodegl/hwconv.h libnodegl/src/nodegl/core/hwconv.h
mv libnodegl/hwupload.c libnodegl/src/nodegl/core/hwupload.c
mv libnodegl/hwupload.h libnodegl/src/nodegl/core/hwupload.h
mv libnodegl/hwupload_common.c libnodegl/src/nodegl/core/hwupload_common.c
mv libnodegl/image.c libnodegl/src/nodegl/core/image.c
mv libnodegl/image.h libnodegl/src/nodegl/core/image.h
mv libnodegl/limits.h libnodegl/src/nodegl/core/limits.h
mv libnodegl/log.c libnodegl/src/nodegl/core/log.c
mv libnodegl/log.h libnodegl/src/nodegl/core/log.h
mv libnodegl/math_utils.c libnodegl/src/nodegl/core/math_utils.c
mv libnodegl/math_utils.h libnodegl/src/nodegl/core/math_utils.h
mv libnodegl/memory.c libnodegl/src/nodegl/core/memory.c
mv libnodegl/memory.h libnodegl/src/nodegl/core/memory.h
mv libnodegl/node_animated.c libnodegl/src/nodegl/core/node_animated.c
mv libnodegl/node_animatedbuffer.c libnodegl/src/nodegl/core/node_animatedbuffer.c
mv libnodegl/node_animkeyframe.c libnodegl/src/nodegl/core/node_animkeyframe.c
mv libnodegl/node_block.c libnodegl/src/nodegl/core/node_block.c
mv libnodegl/node_buffer.c libnodegl/src/nodegl/core/node_buffer.c
mv libnodegl/node_camera.c libnodegl/src/nodegl/core/node_camera.c
mv libnodegl/node_circle.c libnodegl/src/nodegl/core/node_circle.c
mv libnodegl/node_compute.c libnodegl/src/nodegl/core/node_compute.c
mv libnodegl/node_computeprogram.c libnodegl/src/nodegl/core/node_computeprogram.c
mv libnodegl/node_geometry.c libnodegl/src/nodegl/core/node_geometry.c
mv libnodegl/node_graphicconfig.c libnodegl/src/nodegl/core/node_graphicconfig.c
mv libnodegl/node_group.c libnodegl/src/nodegl/core/node_group.c
mv libnodegl/node_hud.c libnodegl/src/nodegl/core/node_hud.c
mv libnodegl/node_identity.c libnodegl/src/nodegl/core/node_identity.c
mv libnodegl/node_io.c libnodegl/src/nodegl/core/node_io.c
mv libnodegl/node_media.c libnodegl/src/nodegl/core/node_media.c
mv libnodegl/node_program.c libnodegl/src/nodegl/core/node_program.c
mv libnodegl/node_quad.c libnodegl/src/nodegl/core/node_quad.c
mv libnodegl/node_render.c libnodegl/src/nodegl/core/node_render.c
mv libnodegl/node_resourceprops.c libnodegl/src/nodegl/core/node_resourceprops.c
mv libnodegl/node_rotate.c libnodegl/src/nodegl/core/node_rotate.c
mv libnodegl/node_rotatequat.c libnodegl/src/nodegl/core/node_rotatequat.c
mv libnodegl/node_rtt.c libnodegl/src/nodegl/core/node_rtt.c
mv libnodegl/node_scale.c libnodegl/src/nodegl/core/node_scale.c
mv libnodegl/node_streamed.c libnodegl/src/nodegl/core/node_streamed.c
mv libnodegl/node_streamedbuffer.c libnodegl/src/nodegl/core/node_streamedbuffer.c
mv libnodegl/node_text.c libnodegl/src/nodegl/core/node_text.c
mv libnodegl/node_texture.c libnodegl/src/nodegl/core/node_texture.c
mv libnodegl/node_time.c libnodegl/src/nodegl/core/node_time.c
mv libnodegl/node_timerangefilter.c libnodegl/src/nodegl/core/node_timerangefilter.c
mv libnodegl/node_timerangemodes.c libnodegl/src/nodegl/core/node_timerangemodes.c
mv libnodegl/node_transform.c libnodegl/src/nodegl/core/node_transform.c
mv libnodegl/node_translate.c libnodegl/src/nodegl/core/node_translate.c
mv libnodegl/node_triangle.c libnodegl/src/nodegl/core/node_triangle.c
mv libnodegl/node_uniform.c libnodegl/src/nodegl/core/node_uniform.c
mv libnodegl/node_userswitch.c libnodegl/src/nodegl/core/node_userswitch.c
mv libnodegl/nodegl.h libnodegl/src/nodegl/core/nodegl.h
mv libnodegl/nodes.c libnodegl/src/nodegl/core/nodes.c
mv libnodegl/nodes.h libnodegl/src/nodegl/core/nodes.h
mv libnodegl/nodes_register.h libnodegl/src/nodegl/core/nodes_register.h
mv libnodegl/params.c libnodegl/src/nodegl/core/params.c
mv libnodegl/params.h libnodegl/src/nodegl/core/params.h
mv libnodegl/pass.c libnodegl/src/nodegl/core/pass.c
mv libnodegl/pass.h libnodegl/src/nodegl/core/pass.h
mv libnodegl/pgcache.c libnodegl/src/nodegl/core/pgcache.c
mv libnodegl/pgcache.h libnodegl/src/nodegl/core/pgcache.h
mv libnodegl/pgcraft.c libnodegl/src/nodegl/core/pgcraft.c
mv libnodegl/pgcraft.h libnodegl/src/nodegl/core/pgcraft.h
mv libnodegl/pipeline.c libnodegl/src/nodegl/core/pipeline.c
mv libnodegl/pipeline.h libnodegl/src/nodegl/core/pipeline.h
mv libnodegl/program.c libnodegl/src/nodegl/core/program.c
mv libnodegl/program.h libnodegl/src/nodegl/core/program.h
mv libnodegl/rendertarget.c libnodegl/src/nodegl/core/rendertarget.c
mv libnodegl/rendertarget.h libnodegl/src/nodegl/core/rendertarget.h
mv libnodegl/rnode.c libnodegl/src/nodegl/core/rnode.c
mv libnodegl/rnode.h libnodegl/src/nodegl/core/rnode.h
mv libnodegl/serialize.c libnodegl/src/nodegl/core/serialize.c
mv libnodegl/texture.c libnodegl/src/nodegl/core/texture.c
mv libnodegl/texture.h libnodegl/src/nodegl/core/texture.h
mv libnodegl/topology.h libnodegl/src/nodegl/core/topology.h
mv libnodegl/transforms.c libnodegl/src/nodegl/core/transforms.c
mv libnodegl/transforms.h libnodegl/src/nodegl/core/transforms.h
mv libnodegl/type.h libnodegl/src/nodegl/core/type.h
mv libnodegl/utils.c libnodegl/src/nodegl/core/utils.c
mv libnodegl/utils.h libnodegl/src/nodegl/core/utils.h
mv libnodegl/android/javb/OnFrameAvailableListener.java libnodegl/src/nodegl/porting/android/OnFrameAvailableListener.java
mv libnodegl/android_ctx.c libnodegl/src/nodegl/porting/android/android_ctx.c
mv libnodegl/android_ctx.h libnodegl/src/nodegl/porting/android/android_ctx.h
mv libnodegl/android_handler.c libnodegl/src/nodegl/porting/android/android_handler.c
mv libnodegl/android_handler.h libnodegl/src/nodegl/porting/android/android_handler.h
mv libnodegl/android_handlerthread.c libnodegl/src/nodegl/porting/android/android_handlerthread.c
mv libnodegl/android_handlerthread.h libnodegl/src/nodegl/porting/android/android_handlerthread.h
mv libnodegl/android_imagereader.c libnodegl/src/nodegl/porting/android/android_imagereader.c
mv libnodegl/android_imagereader.h libnodegl/src/nodegl/porting/android/android_imagereader.h
mv libnodegl/android_looper.c libnodegl/src/nodegl/porting/android/android_looper.c
mv libnodegl/android_looper.h libnodegl/src/nodegl/porting/android/android_looper.h
mv libnodegl/android_surface.c libnodegl/src/nodegl/porting/android/android_surface.c
mv libnodegl/android_surface.h libnodegl/src/nodegl/porting/android/android_surface.h
mv libnodegl/android_utils.c libnodegl/src/nodegl/porting/android/android_utils.c
mv libnodegl/android_utils.h libnodegl/src/nodegl/porting/android/android_utils.h
mv libnodegl/jni_utils.c libnodegl/src/nodegl/porting/android/jni_utils.c
mv libnodegl/jni_utils.h libnodegl/src/nodegl/porting/android/jni_utils.h
mv libnodegl/glcontext_eagl.m libnodegl/src/nodegl/porting/eagl/glcontext_eagl.m
mv libnodegl/glcontext_egl.c libnodegl/src/nodegl/porting/egl/glcontext_egl.c
mv libnodegl/buffer_gl.c libnodegl/src/nodegl/porting/gl/buffer_gl.c
mv libnodegl/buffer_gl.h libnodegl/src/nodegl/porting/gl/buffer_gl.h
mv libnodegl/format_gl.c libnodegl/src/nodegl/porting/gl/format_gl.c
mv libnodegl/format_gl.h libnodegl/src/nodegl/porting/gl/format_gl.h
mv libnodegl/gctx_gl.c libnodegl/src/nodegl/porting/gl/gctx_gl.c
mv libnodegl/gctx_gl.h libnodegl/src/nodegl/porting/gl/gctx_gl.h
mv libnodegl/glcontext.c libnodegl/src/nodegl/porting/gl/glcontext.c
mv libnodegl/glcontext.h libnodegl/src/nodegl/porting/gl/glcontext.h
mv libnodegl/gldefinitions_data.h libnodegl/src/nodegl/porting/gl/gldefinitions_data.h
mv libnodegl/glfeatures_data.h libnodegl/src/nodegl/porting/gl/glfeatures_data.h
mv libnodegl/glfunctions.h libnodegl/src/nodegl/porting/gl/glfunctions.h
mv libnodegl/glincludes.h libnodegl/src/nodegl/porting/gl/glincludes.h
mv libnodegl/glstate.c libnodegl/src/nodegl/porting/gl/glstate.c
mv libnodegl/glstate.h libnodegl/src/nodegl/porting/gl/glstate.h
mv libnodegl/glwrappers.h libnodegl/src/nodegl/porting/gl/glwrappers.h
mv libnodegl/gtimer_gl.c libnodegl/src/nodegl/porting/gl/gtimer_gl.c
mv libnodegl/gtimer_gl.h libnodegl/src/nodegl/porting/gl/gtimer_gl.h
mv libnodegl/pipeline_gl.c libnodegl/src/nodegl/porting/gl/pipeline_gl.c
mv libnodegl/pipeline_gl.h libnodegl/src/nodegl/porting/gl/pipeline_gl.h
mv libnodegl/program_gl.c libnodegl/src/nodegl/porting/gl/program_gl.c
mv libnodegl/program_gl.h libnodegl/src/nodegl/porting/gl/program_gl.h
mv libnodegl/rendertarget_gl.c libnodegl/src/nodegl/porting/gl/rendertarget_gl.c
mv libnodegl/rendertarget_gl.h libnodegl/src/nodegl/porting/gl/rendertarget_gl.h
mv libnodegl/texture_gl.c libnodegl/src/nodegl/porting/gl/texture_gl.c
mv libnodegl/texture_gl.h libnodegl/src/nodegl/porting/gl/texture_gl.h
mv libnodegl/topology_gl.c libnodegl/src/nodegl/porting/gl/topology_gl.c
mv libnodegl/topology_gl.h libnodegl/src/nodegl/porting/gl/topology_gl.h
mv libnodegl/type_gl.c libnodegl/src/nodegl/porting/gl/type_gl.c
mv libnodegl/type_gl.h libnodegl/src/nodegl/porting/gl/type_gl.h
mv libnodegl/hwupload_mediacodec_gl.c libnodegl/src/nodegl/porting/hwupload/hwupload_mediacodec_gl.c
mv libnodegl/hwupload_vaapi_gl.c libnodegl/src/nodegl/porting/hwupload/hwupload_vaapi_gl.c
mv libnodegl/hwupload_videotoolbox_darwin_gl.c libnodegl/src/nodegl/porting/hwupload/hwupload_videotoolbox_darwin_gl.c
mv libnodegl/hwupload_videotoolbox_darwin_vk.m libnodegl/src/nodegl/porting/hwupload/hwupload_videotoolbox_darwin_vk.m
mv libnodegl/hwupload_videotoolbox_ios_gl.c libnodegl/src/nodegl/porting/hwupload/hwupload_videotoolbox_ios_gl.c
mv libnodegl/glcontext_nsgl.m libnodegl/src/nodegl/porting/nsgl/glcontext_nsgl.m
mv libnodegl/vaapi.c libnodegl/src/nodegl/porting/vaapi/vaapi.c
mv libnodegl/vaapi.h libnodegl/src/nodegl/porting/vaapi/vaapi.h
mv libnodegl/buffer_vk.c libnodegl/src/nodegl/porting/vk/buffer_vk.c
mv libnodegl/buffer_vk.h libnodegl/src/nodegl/porting/vk/buffer_vk.h
mv libnodegl/format_vk.c libnodegl/src/nodegl/porting/vk/format_vk.c
mv libnodegl/format_vk.h libnodegl/src/nodegl/porting/vk/format_vk.h
mv libnodegl/gctx_vk.c libnodegl/src/nodegl/porting/vk/gctx_vk.c
mv libnodegl/gctx_vk.h libnodegl/src/nodegl/porting/vk/gctx_vk.h
mv libnodegl/gtimer_vk.c libnodegl/src/nodegl/porting/vk/gtimer_vk.c
mv libnodegl/gtimer_vk.h libnodegl/src/nodegl/porting/vk/gtimer_vk.h
mv libnodegl/pipeline_vk.c libnodegl/src/nodegl/porting/vk/pipeline_vk.c
mv libnodegl/pipeline_vk.h libnodegl/src/nodegl/porting/vk/pipeline_vk.h
mv libnodegl/program_vk.c libnodegl/src/nodegl/porting/vk/program_vk.c
mv libnodegl/program_vk.h libnodegl/src/nodegl/porting/vk/program_vk.h
mv libnodegl/rendertarget_vk.c libnodegl/src/nodegl/porting/vk/rendertarget_vk.c
mv libnodegl/rendertarget_vk.h libnodegl/src/nodegl/porting/vk/rendertarget_vk.h
mv libnodegl/texture_vk.c libnodegl/src/nodegl/porting/vk/texture_vk.c
mv libnodegl/texture_vk.h libnodegl/src/nodegl/porting/vk/texture_vk.h
mv libnodegl/topology_vk.c libnodegl/src/nodegl/porting/vk/topology_vk.c
mv libnodegl/topology_vk.h libnodegl/src/nodegl/porting/vk/topology_vk.h
mv libnodegl/vkcontext.c libnodegl/src/nodegl/porting/vk/vkcontext.c
mv libnodegl/vkcontext.h libnodegl/src/nodegl/porting/vk/vkcontext.h
mv libnodegl/glcontext_wgl.c libnodegl/src/nodegl/porting/wgl/glcontext_wgl.c
mv libnodegl/test_asm.c libnodegl/test/test_asm.c
mv libnodegl/test_colorconv.c libnodegl/test/test_colorconv.c
mv libnodegl/test_darray.c libnodegl/test/test_darray.c
mv libnodegl/test_draw.c libnodegl/test/test_draw.c
mv libnodegl/test_hmap.c libnodegl/test/test_hmap.c
mv libnodegl/test_utils.c libnodegl/test/test_utils.c
mv libnodegl/gen_doc.c libnodegl/tools/gen_doc/gen_doc.c
mv libnodegl/gen_specs.c libnodegl/tools/gen_specs/gen_specs.c

**Comment by jmoguillansky-gpsw** [comment]
ngfx-integration code rebased on latest Vulkan branch
